src/openzeppelin/governance/timelock/library.cairo

```diff
@@ -0,0 +1,505 @@
+
+        let (timestamp: felt) = get_timestamp(id)
+        let (pending: felt) = is_le(DONE_TIMESTAMP + 1, timestamp)
+
+        return (pending=pending)
+    end
+
+    func is_operation_ready{
+        syscall_ptr: felt*, 
+        pedersen_ptr: HashBuiltin*, 
+        range_check_ptr
+    }(id: felt) -> (ready: felt):
+        alloc_locals
+
+        let (timestamp: felt) = get_timestamp(id)
+        let (block_timestamp: felt) = get_block_timestamp()
+
+        let (pending: felt) = is_operation_pending(id)
+        let (ready: felt) = is_le(timestamp, block_timestamp)
+
+        if pending + ready == 2:
+            return (TRUE)
+        else:
+            return (FALSE)
+        end
+    end
+
+    func is_operation_done{
+        syscall_ptr: felt*, 
+        pedersen_ptr: HashBuiltin*, 
+        range_check_ptr
+    }(id: felt) -> (done: felt):
+        let (timestamp: felt) = get_timestamp(id)
+
+        if timestamp == DONE_TIMESTAMP:
+            return (TRUE)
+        else:
+            return (FALSE)
+        end
+    end
+
+    func update_delay{
+        syscall_ptr: felt*, 
+        pedersen_ptr: HashBuiltin*, 
+        range_check_ptr
+    }(min_delay: felt):
+        let (old_min_delay: felt) = Timelock_min_delay.read()
+        Timelock_min_delay.write(min_delay)
+        MinDelayChange.emit(old_min_delay, min_delay)
+
+        return ()
+    end
+
+    func schedule{
+        syscall_ptr: felt*, 
+        pedersen_ptr: HashBuiltin*, 
+        range_check_ptr
+    }(
+        call_array_len: felt,
+        call_array: TimelockCall*,
+        calldata_len: felt,
+        calldata: felt*,
+        predecessor: felt,
+        salt: felt,
+        delay: felt,
+    ):
+        alloc_locals
+
+        let (id: felt) = hash_operation(call_array_len, call_array, calldata, predecessor, salt)
+
+        let (calls: Call*) = alloc()
+        _from_timelock_calls_to_calls(call_array_len, call_array, calldata, calls)
+        _schedule(id, delay, call_array_len, calls, predecessor)
+
+        return ()
+    end
+
+    func execute{
+        syscall_ptr: felt*, 
+        pedersen_ptr: HashBuiltin*, 
+        range_check_ptr
+    }(
+        call_array_len: felt,
+        call_array: TimelockCall*,
+        calldata_len: felt,
+        calldata: felt*,
+        predecessor: felt,
+        salt: felt,
+    ):
+        alloc_locals
+
+        let (id: felt) = hash_operation(call_array_len, call_array, calldata, predecessor, salt)
+        let (ready: felt) = is_operation_ready(id)
+
+        with_attr error_message("Timelock: operation is not ready"):
+            assert ready = TRUE
+        end
+
+        if predecessor != 0:
+            let (predecessor_done: felt) = is_operation_done(predecessor)
+
+            with_attr error_message("Timelock: missing dependency"):
+                assert predecessor_done = TRUE
+            end
+        end
+
+        _execute(id, call_array_len, call_array, calldata_len, calldata, predecessor)
```

Comment:
as with `_schedule`, I'm not convinced there's any benefit in abstracting it out. Can you think of any?